Timeline Component! 🔮

dmarcey opened the pull request:
Adds a Timeline component

Closes #677

### Screenshots
![image](https://user-images.githubusercontent.com/5487287/75356449-fb63cf00-587d-11ea-93d0-994db974a030.png)


### Merge checklist
- [x] Added or updated TypeScript definitions (`index.d.ts`) if necessary
- [x] Added/updated tests
- [x] Added/updated documentation
- [ ] Tested in Chrome
- [ ] Tested in Firefox
- [ ] Tested in Safari
- [ ] Tested in Edge

BinaryMuse commented on docs/content/Timeline.md:
```diff
@@ -0,0 +1,138 @@
+
+## Component props
+
+### Timeline.Item
```
In the code, this is a prop on `Timeline` and not `Timeline.Item`.

```suggestion
### Timeline
```

BinaryMuse commented on src/Timeline.js:
```diff
@@ -0,0 +1,165 @@
+  children: PropTypes.node,
+  clipSidebar: PropTypes.bool,
+  theme: PropTypes.object,
+  ...COMMON.propTypes
```
The TypeScript definitions extend `FlexProps`, but `FLEX` is not included in any of the `propTypes` of the components that wrap a `Flex`, nor is `${FLEX}` included in items that don't wrap a `Flex`. Also, things like `Timeline.Body` which is a wrapped `Box` should inherit from Box's props instead of `TextProps` as it currently does.

dmarcey replied:
Thanks! I'll add the ${FLEX} tokens to the `Timeline` and `Timeline.Item` styles, as well as updating the `propTypes` for both!

dmarcey replied:
Also fixed the TS interfaces for the `Timeline.Body` and `Timeline.Break`

BinaryMuse commented on src/Timeline.js:
```diff
@@ -0,0 +1,165 @@
+
+Timeline.Item = props => {
+  return <TimelineItemInternal {...props}>{props.children}</TimelineItemInternal>
+}
```
What's the purpose of the additional indirection here?

dmarcey replied:
Good catch! There isn't any purpose any more, originally I was doing something in here, adding an extra className, maybe? But I must have refactored it all away! Will update to make the styled component be the `Timeline.Item` directly.

dmarcey commented:
Thanks for the diff, @BinaryMuse! That was super helpful - I switched to using the `css` helper to get rid of the `props => props.theme` syntax and updated the `propTypes`.

I don't think that I am able to get rid of the `${COMMON} props from `TimelineBadgeInternal` and `Timeline.Item`, however. When I tried to do that, I noticed that the common system props (like `bg` or `pt)` stopped working. Does it make sense that I would need to continue to include those?

emplums commented:
@dmarcey I did some digging and figured out the reason that you need to continue to include them is because the css that you add to `TimelineBadgeInteral` in styled-components is applied _after_ the CSS created by using one of the ${COMMON} system props. That's also why we always tend to list the system props last in the styled components function, so it always overrides the custom styles included in the component. As you can see, it gets pretty convoluted when extending a component like we're doing here. The order of styles applied in this case would be:

- all custom styles from the `Flex` component
- system prop styles from the `Flex` component
- all custom styles from `TimelineBadgeInternal` component
- If you were to add ${COMMON} to the end of the `TimelineBadgeInteral` styles, they would be applied at this point (last). you'd see a duplication of styles generated (once for the `COMMON` implementation in `Flex` and once for the `COMMON` implementation in `TimelineBadgeInternal`

Here's how that looks in the styles produced:
![image](https://user-images.githubusercontent.com/8960591/76468358-09027400-63a9-11ea-9098-d329b1955dcc.png)

One solution for this particular case is to set the default value for the `bg` prop for `Timeline.Badge` to `gray.2` instead of applying it in the custom CSS. That will make sure that the background is gray by default, but if someone provides a custom background it gets handled by the `Flex` component without getting overridden by custom CSS in `TimelineBadgeInternal`

This has been something lingering in the back of my mind regarding extending components from other components 😬 It sort of breaks down whenever you have "default" styles that override system styles like we do for padding and background color here. We could opt to never extend components within the library to avoid this. Curious if you have any ideas!

emplums commented:
Ok messed around a bit with this today and I think this should work as an alternative! :) I moved most of the styles in the styled components function into the `Flex` component as props. That way they can reliably be overridden if a user provides the same props.

```
const TimelineBadgeBase = (props) => {
  return (
    <Flex flexShrink={0} className={classnames(props.className, 'TimelineItem-Badge')} color='gray.7' width='32px' height='32px' mr={2} ml='-15px' alignItems='center' justifyContent='center' {...props}>
      {props.children}
    </Flex>
  )
}

Timeline.Badge = styled(TimelineBadgeBase)`
  position: relative;
  overflow: hidden;
  z-index: 1;
  border: 2px solid ${get('colors.white')};
  border-radius: 50%;
`
```

You could take this even further and completely remove the need for using any custom styles by doing this:

```
Timeline.Badge = (props) => {
  return (
    <Relative zIndex={1}>
      <Flex flexShrink={0} className={classnames(props.className, 'TimelineItem-Badge')} css={`border-radius: 50%; border: 2px solid ${get('colors.white')}`} overflow='hidden' color='gray.7' width='32px' height='32px' mr={2} ml='-15px' alignItems='center' justifyContent='center' {...props}>
        {props.children}
      </Flex>
    </Relative>
  )
}

```

The amount of props used in the `Flex` component feels a little gnarly but not sure if that should stop us from shipping something like that. What do you think?

emplums commented:
This could lead to a recommendation for designing components/extending components: "Any styles that correspond to styled-system props should be applied via props instead of custom styles to ensure that styled-system props are applied properly in the consuming application"
